lib/shared/s3-packages.js

```diff
@@ -0,0 +1,243 @@
+    return exports.BUCKET_URL.SNAPSHOT;
+  }
+
+  return null;
+};
+
+/**
+ * @summary Get all remote versions from an S3 bucket
+ * @function
+ * @private
+ *
+ * @description
+ * We memoize based on the assumption that the received latest version
+ * number will not increase while the application is running.
+ *
+ * @param {String} bucketUrl - s3 bucket url
+ * @fulfil {String[]} - remote versions
+ * @returns {Promise}
+ *
+ * @example
+ * s3Packages.getRemoteVersions(s3Packages.BUCKET_URL.PRODUCTION).then((versions) => {
+ *   _.each(versions, (version) => {
+ *     console.log(version);
+ *   });
+ * });
+ */
+exports.getRemoteVersions = _.memoize((bucketUrl) => {
+  if (_.isNil(bucketUrl)) {
+    return Bluebird.reject(new Error(`Invalid bucket url: ${bucketUrl}`));
+  }
+
+  /* eslint-disable lodash/prefer-lodash-method */
+
+  return request.getAsync(bucketUrl)
+
+  /* eslint-enable lodash/prefer-lodash-method */
+
+    .get('body')
+    .then(xml.parseStringAsync)
+    .get('ListBucketResult')
+    .then((bucketResult) => {
+      return _.get(bucketResult, [ 'Contents' ], []);
+    })
+    .reduce((accumulator, entry) => {
+      const [ name, version ] = _.split(_.first(entry.Key), '/');
+
+      if (name === S3_PACKAGE_NAME) {
+        if (_.isNil(accumulator[version])) {
+          accumulator[version] = 1;
+        } else {
+          accumulator[version] += 1;
+        }
+      }
+
+      return accumulator;
+    }, [])
+    .then((versions) => {
+      return _.keys(_.pickBy(versions, (occurrences) => {
+        return occurrences >= NUMBER_OF_PACKAGES;
+      }));
+    })
+    .catch({
+      code: 'ENOTFOUND'
+    }, {
+      code: 'ETIMEDOUT'
+    }, () => {
+      return [];
+    });
+});
+
+/**
+ * @summary Get the latest available version from an S3 bucket
+ * @function
+ * @public
+ *
+ * @param {String} bucketUrl - an s3 bucket url
+ * @param {Object} options - options
+ * @param {String} [options.range] - semver range
+ * @fulfil {String} - latest version
+ * @returns {Promise}
+ *
+ * @example
+ * s3Packages.getLatestVersion(s3Packages.BUCKET_URL.PRODUCTION, {
+ *   range: '>=2.0.0'
+ * }).then((latestVersion) => {
+ *   console.log(`The latest version is: ${latestVersion}`);
+ * });
+ */
+exports.getLatestVersion = (bucketUrl, options = {}) => {
+
+  // For manual testing purposes
+  const ETCHER_FAKE_S3_LATEST_VERSION = process.env.ETCHER_FAKE_S3_LATEST_VERSION;
+  if (semver.valid(ETCHER_FAKE_S3_LATEST_VERSION)) {
+    return Bluebird.resolve(ETCHER_FAKE_S3_LATEST_VERSION);
+  }
+
+  /* eslint-disable lodash/prefer-lodash-method */
+
+  return exports.getRemoteVersions(bucketUrl).filter((version) => {
+
+  /* eslint-enable lodash/prefer-lodash-method */
+
+    // These checks allow us to ignore snapshot builds in production
+    // buckets, and viceversa, which could have been uploaded by mistake.
+
+    const releaseType = release.getReleaseType(version);
+
+    if (bucketUrl === exports.BUCKET_URL.PRODUCTION
```

Comment:
I guess the issue is that we could pass a snapshot URL, but tell the function to only consider production builds, which would make no sense. A better approach seems to be making this function accept a release type instead of a URL, and internally calculate the correct URL.